### PR TITLE
[BEAM-9633] Add PubsubIO performance test

### PR DESCRIPTION
**Pull request by piotr-szuberski**
Add PubsubIO performance test

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by piotr-szuberski** [comment]
@kamilwu

**Comment by kamilwu** [comment]
Retest this please

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    }
+
+    args = self.pipeline.get_full_options_as_args(**self.extra_opts)
+
+    parser = argparse.ArgumentParser()
+    _, pipeline_args = parser.parse_known_args(args)
+
+    pipeline_options = PipelineOptions(pipeline_args)
+
+    self.pipeline = TestPipeline(options=pipeline_options)
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py:152]: You don't have to create PipelineOptions and ArgumentParser objects. This should be enough:
```
args = self.pipeline.get_full_options_as_args(**extra_opts)
self.pipeline = TestPipeline(argv=args)
```

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+    )
+
+    self.extra_opts = {
+        'on_success_matcher': all_of(pubsub_msg_verifier),
+        'wait_until_finish_duration': self.timeout * 1000,
+        'streaming': True,
+        'save_main_session': True
+    }
+
+    args = self.pipeline.get_full_options_as_args(**self.extra_opts)
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py:145]: I think since we don't use `self.extra_opts` anywhere else, the assignment to `self` is redundant.

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+#
+
+"""
+Performance PubsubIO streaming test for Read operation.
+
+Example for TestDataflowRunner:
+
+python -m apache_beam.io.gcp.pubsub_read_perf_test \
+    --test-pipeline-options="
+    --runner=TestDataflowRunner
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py:25]: I would put emphasis on that the runner must be `TestDataflowRunner`, not `DataflowRunner`. This is very important information for anyone who would run this code.

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+  def _setup_env(self):
+    if not self.pipeline.get_option('timeout'):
+      logging.error('--timeout argument is required.')
+      sys.exit(1)
+
+    self.num_of_messages = int(self.input_options.get('num_records'))
+    self.timeout = int(self.pipeline.get_option('timeout'))
+
+  def _setup_pubsub(self):
+    def init_input(number_of_messages):
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py:98]: If you don't use a closure, I think you can make this function a top-level one. It would be easier to understand the program flow.

File: sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+    class ToPubsubMessage(beam.DoFn):
+      counter = 0
+
+      def process(self, element):
+        self.counter += 1
+        from apache_beam.io import WriteToPubSub, PubsubMessage
+        message = PubsubMessage(
+          data=element[1],
+          attributes={'ack_id': bytes(str(self.counter), 'utf-8')})
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py:87]: The same incompatibility error I mentioned earlier.

File: sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+  def __init__(self):
+    super(PubsubWritePerfTest, self).__init__()
+    self._setup_env()
+    self._setup_pubsub()
+    self._setup_pipeline()
+
+  def test(self):
+
+    class ToPubsubMessage(beam.DoFn):
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py:79]: For trivial `DoFn`'s, you can also define a simple function instead of `DoFn`. Then, if it's a generator function, you can use a `beam.FlatMap` in the pipeline. Here are some examples: https://beam.apache.org/documentation/transforms/python/elementwise/flatmap/

**Comment by kamilwu** [comment]
One more comment: As I see, a lot of code is duplicated. Do you think you can reduce the code duplication in some way?

**Comment by piotr-szuberski** [comment]
> One more comment: As I see, a lot of code is duplicated. Do you think you can reduce the code duplication in some way?

There is a lot of duplication, but I thought that in case of tests it's not a problem. It can easily be done with making a superclass with most of the setup.

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+import argparse
+import logging
+import sys
+import uuid
+
+from hamcrest import all_of
+
+import apache_beam as beam
+from apache_beam.io import ReadFromPubSub, WriteToPubSub, PubsubMessage
```
[the remark this answers is not in the document]
Comment (piotr-szuberski) [reply sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py]: Done

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+  def _setup_pubsub(self):
+    def init_input(number_of_messages):
+      for n in range(number_of_messages):
+        message = PubsubMessage(
+          bytes(str(n), 'utf-8'),
+          {'ack_id': bytes(str(n), 'utf-8')})
+        bytes_message = WriteToPubSub.to_proto_str(message)
+        self.pub_client.publish(self.topic.name, bytes_message)
+
+    from google.cloud import pubsub
```
[the remark this answers is not in the document]
Comment (piotr-szuberski) [reply sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py:106]: It was done like that in other pubsub examples. But I think it's a better idea to do as you suggest.

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py (hunk 1 above)
Comment (piotr-szuberski) [reply sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py:152]: Done

File: sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py (hunk 5 above)
Comment (piotr-szuberski) [reply sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py:87]: Done

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py (hunk 4 above)
Comment (piotr-szuberski) [reply sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py:98]: @kamilwu you mean the init input function? Would it be preferable to start it with underscore to indicate it shouldn't be used outside of the class?

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py (hunk 3 above)
Comment (piotr-szuberski) [reply sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py:25]: Done

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+      sys.exit(1)
+
+    self.num_of_messages = int(self.input_options.get('num_records'))
+    self.timeout = int(self.pipeline.get_option('timeout'))
+
+  def _setup_pubsub(self):
+    def init_input(number_of_messages):
+      for n in range(number_of_messages):
+        message = PubsubMessage(
+          bytes(str(n), 'utf-8'),
```
[the remark this answers is not in the document]
Comment (piotr-szuberski) [reply sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py]: Done

File: sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py (hunk 6 above)
Comment (piotr-szuberski) [reply sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py:79]: I'm not sure whether I'm able to assign an incrementing id to the elements using FlatMap. The point of using DoFn was to hold a counter as an id generator to ensure its uniqueness.

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py (hunk 4 above)
Comment (kamilwu) [reply sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py]: Yes, I meant `init_input` function. Let's add an underscore.

Comment (kamilwu) [reply sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py]: What about changing the name of this function so it would be more descriptive? What do you think about `_inject_input_data`?

File: sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py (hunk 6 above)
Comment (kamilwu) [reply sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py:70]: Oh yes, I overlooked that variable. Let's keep as it is.

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py (hunk 4 above)
Comment (piotr-szuberski) [reply sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py]: Good idea!

File: .test-infra/jenkins/job_PerformanceTests_PubsubIO_Python.groovy [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                autoscaling_algorithm: 'NONE',  // Disable autoscale the worker pool.
+                timeout              : 3600,
+        ]
+]
+
+def executeJob = { scope, testConfig ->
+    commonJobProperties.setTopLevelMainJobProperties(scope, 'master', 240)
+
+    loadTestsBuilder.loadTest(scope, testConfig.title, testConfig.runner,
+            CommonTestProperties.SDK.PYTHON, testConfig.pipelineOptions, testConfig.test)
```
Comment (kamilwu) [inline .test-infra/jenkins/job_PerformanceTests_PubsubIO_Python.groovy:66]: Recently, we've migrated Load Tests and BigQuery perf tests to Python 3.7. It would be good to keep using Python 3.7 in PubSub perf tests as well.

File: sdks/python/apache_beam/io/gcp/tests/pubsub_matcher.py [hunk 11]
```diff
@@ -38,7 +38,6 @@
   pubsub = None
 
 DEFAULT_TIMEOUT = 5 * 60
-MAX_MESSAGES_IN_ONE_PULL = 50
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/tests/pubsub_matcher.py:41]: We can rename this constant to `DEFAULT_MAX_MESSAGES_IN_ONE_PULL` and use as a default parameter value.

File: sdks/python/apache_beam/io/gcp/tests/pubsub_matcher.py [hunk 12]
```diff
@@ -57,7 +56,9 @@ def __init__(
       expected_msg_len=None,
       timeout=DEFAULT_TIMEOUT,
       with_attributes=False,
-      strip_attributes=None):
+      strip_attributes=None,
+      sleep_time=1,
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/tests/pubsub_matcher.py:60]: I'm fine with additional parameters, but we need to update a docstring.

**Comment by kamilwu** [comment]
Thanks for extracting a superclass. It looks much better now.

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Seed Job

File: sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+          sub_name,
+          self.topic_name,
+          ack_deadline_seconds=300,
+          timeout=1200,
+          retain_acked_messages=True,
+      )
+
+    _ = self.pub_client.create_topic(self.topic_name)
+    make_subscription(self.write_matcher_sub_name)
+    make_subscription(self.read_sub_name)
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py:98]: How about creating a subscription for reading messages in the read test? I'm pretty sure the write test does not need it.

File: sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+    make_subscription(self.read_matcher_sub_name)
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py:99]: The same situation as above.

File: sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py [hunk 15]
```diff
@@ -0,0 +1,104 @@
...
+      _ = self.sub_client.create_subscription(
+          sub_name,
+          self.topic_name,
+          ack_deadline_seconds=300,
+          timeout=1200,
+          retain_acked_messages=True,
+      )
+
+    _ = self.pub_client.create_topic(self.topic_name)
+    make_subscription(self.write_matcher_sub_name)
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py:97]: Can we shift the responsibility of cleaning up this subscription onto this test? Remember that both tests can be executed independently, without Jenkins.

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+"""
+Performance PubsubIO streaming test for Read operation.
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py:19]: It would be good to notice that the input topic must already exist before executing the test.

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    --project=<GCP_PROJECT_ID>
+    --temp_location=gs://<BUCKET_NAME>/tmp
+    --staging_location=gs://<BUCKET_NAME>/staging
+    --timeout=<TIME_IN_SECONDS>
+    --pubsub_name=<PUBSUB_NAME_TO_BE_CREATED>
+    --publish_to_big_query=<OPTIONAL><true/false>
+    --metrics_dataset=<OPTIONAL>
+    --metrics_table=<OPTIONAL>
+    --dataflow_worker_jar=<OPTIONAL>
+    --input_options='{
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py:38]: This parameter is not used. I think we can remove this. Please also remove the code that checks if --input_option is present at the base class.

File: .test-infra/jenkins/job_PerformanceTests_PubsubIO_Python.groovy [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+]
+
+def executeJob = { scope, testConfig ->
+    commonJobProperties.setTopLevelMainJobProperties(scope, 'master', 240)
+
+    loadTestsBuilder.loadTest(scope, testConfig.title, testConfig.runner,
+            CommonTestProperties.SDK.PYTHON_37, testConfig.pipelineOptions, testConfig.test)
+}
+
+PhraseTriggeringPostCommitBuilder.postCommitJob(
```
Comment (kamilwu) [inline .test-infra/jenkins/job_PerformanceTests_PubsubIO_Python.groovy:82]: I like the idea of reusing output topic for the purpose of the read test. But, in that case, you have to make sure that the write test is executed first. At the moment, both tests start independently at the same hour ('H 15 * * *'), which can cause some issues.

**Comment by kamilwu** [comment]
Thanks @piotr-szuberski. I've left some comments. I feel we are already very close.

File: sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py (hunk 13 above)
Comment (piotr-szuberski) [reply sdks/python/apache_beam/io/gcp/pubsub_write_perf_test.py:98]: It is needed because the subscription must exists when the messages are being published. If I move it to the read test then the subscription will be empty.

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py (hunk 16 above)
Comment (piotr-szuberski) [reply sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py:19]: Sure, but still the read subscriptions can't be cleaned here so overall I would suggest making  one write-read test instead of splitting them

File: sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py (hunk 17 above)
Comment (piotr-szuberski) [reply sdks/python/apache_beam/io/gcp/pubsub_read_perf_test.py:38]: input options is used for the synthetic source in write, it specifies the number of messages and the size of each message. It can be not used and some num_of_messages and length_of_messages can be used instead, but it can as well stay as it is

File: .test-infra/jenkins/job_PerformanceTests_PubsubIO_Python.groovy (hunk 18 above)
Comment (piotr-szuberski) [reply .test-infra/jenkins/job_PerformanceTests_PubsubIO_Python.groovy:82]: I moved both tests to one file so now it'll be impossible to run them separately. They were so tightly bounded that it made no sense to run them individually

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run PubsubIO  Performance Test Python

File: .test-infra/jenkins/job_PerformanceTests_PubsubIO_Python.groovy [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+        'beam_PubsubIO_Performance_Test_Python',
+        'Run PubsubIO  Performance Test Python',
```
Comment (kamilwu) [inline .test-infra/jenkins/job_PerformanceTests_PubsubIO_Python.groovy:63]: There are two spaces between `PubsubIO` and `Performance`. Could you fix that?

File: sdks/python/apache_beam/io/gcp/pubsub_io_perf_test.py [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+        logging.error('--%s argument is required.', name)
+        sys.exit(1)
+
+    check_option('timeout')
+    check_option('pubsub_name')
+
+    self.num_of_messages = int(self.input_options.get('num_records'))
+    self.size_of_messages = int(self.input_options.get('value_size'))
+    self.timeout = int(self.pipeline.get_option('timeout'))
+    self.pubsub_name = self.pipeline.get_option('pubsub_name')
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_io_perf_test.py:85]: It'd good to put emphasis on _what_ is being created with this name. I suggest renaming this option to `topic_name`.

File: sdks/python/apache_beam/io/gcp/pubsub_io_perf_test.py [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+
+    _ = self.pub_client.create_topic(self.topic_name)
+    make_subscription(self.write_matcher_sub_name)
+    make_subscription(self.read_sub_name)
+    make_subscription(self.read_matcher_sub_name)
+
+
+class PubsubReadPerfTest(PubsubIOPerfTest):
+  def __init__(self):
+    super(PubsubReadPerfTest, self).__init__('_read')
```
Comment (kamilwu) [inline sdks/python/apache_beam/io/gcp/pubsub_io_perf_test.py:171]: It's totally fine to write all metrics to the same BigQuery table. That's how all IO IT tests in Java SDK works, where we have two types of metric: "read_time" and "write_time". The only thing you have to do is to change the argument that `MeasureTime`'s constructor takes (it's self.metrics_namespace now, which is BigQuery table, but you wanna change that to, say, 'read_time' in PubsubReadPerfTest and 'write_time' in PubsubWritePerfTest)

Comment (kamilwu) [reply sdks/python/apache_beam/io/gcp/pubsub_io_perf_test.py:171]: Thanks to that, your changes to the base class will be unnecessary.

**Comment by kamilwu** [comment]
Run PubsubIO Performance Test Python

**Comment by kamilwu** [comment]
Run PubsubIO Performance Test Python

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Run PubsubIO Performance Test Python

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run PubsubIO Performance Test Python

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run PubsubIO Performance Test Python

**Comment by kamilwu** [comment]
Run PubsubIO Performance Test Python

File: .test-infra/jenkins/job_PerformanceTests_PubsubIO_Python.groovy [hunk 22]
```diff
@@ -41,7 +42,7 @@ def psio_test = [
                 metrics_dataset      : 'beam_performance',
                 metrics_table        : 'psio_io_2GB_msg_results',
                 input_options        : '\'{' +
-                        '"num_records": 2097152,' +
+                        '"num_records": 2097152' +
```
Comment (kamilwu) [inline .test-infra/jenkins/job_PerformanceTests_PubsubIO_Python.groovy:45]: Why did you remove that comma?

Comment (piotr-szuberski) [reply .test-infra/jenkins/job_PerformanceTests_PubsubIO_Python.groovy:45]: It sounds stupid but my cat went through my keyboard and I deleted too many digits and overlooked it

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Run seed job

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Run PubsubIO Performance Test Python

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run PubsubIO Performance Test Python

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run PubsubIO Performance Test Python

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run PubsubIO Performance Test Python

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run PubsubIO Performance Test Python

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run Seed Job

**Comment by kamilwu** [comment]
Run PubsubIO Performance Test Python

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Retest this please

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
Run Python PreCommit

**Comment by kamilwu** [comment]
LGTM, thanks for contribution!
